### PR TITLE
Add Index Permission Panel to role editing page.

### PR DESCRIPTION
**Pull request by yuxitang-amzn**
*Issue #, if available:*

*Description of changes:*
Add Index Permission Panel to role editing page. This is a redo for miss-clicked #196 .

*Out of scope:*
* Handle button click for `Browse and select` and `Create Action Groups` buttons
* Action group preview when panel collapsed

Screenshots:
| Mockup  |  Implementation |
|---|---|
|  ss_mock_role_edit_idxperm_new |  ss_impl_role_edit_idxperm_new |
|  ss_mock_role_edit_idxperm_edit | ss_impl_role_edit_idxperm_edit  |
|  ss_mock_role_edit_idxperm_collapsed | ss_impl_role_edit_idxperm_collapsed |

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: public/apps/configuration/panels/role-edit.tsx [hunk 1]
```diff
@@ -14,26 +14,41 @@
  */
 
 import {
+  EuiAccordion,
```
Comment (tianleh) [inline public/apps/configuration/panels/role-edit.tsx:17]: With your current change, this file is close to 500 lines with 3 major components, cluster permissions, index permissions and tenants. (the name one is relatively simple). The file has mixed constant variables, interfaces and util functions which are only specific to one component. This makes it difficult to mange. When I read it, I have to use cmd+f on the webapge to keep searching where those items are being used. I feel that this is a proper timing to extract major components out to their own files. If you want to do it in stages, probably start with the new code you added.

File: public/apps/configuration/utils/array-state-utils.tsx [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+/**
+ * Update an element in an sub array
+ * @param setStateCallback setState function
+ * @param path lodash path, e.g. [0, indexPattern] or "[0].indexPattern" 
+ * @param newValue value to be updated
+ *
+ * e.g. 
+ *  currentState = [{a: 1}, {a: 2}]
+ *  path = [1, 'a']
```
Comment (tianleh) [inline public/apps/configuration/utils/array-state-utils.tsx:33]: Minor: the index 1 causes some confusion with the value 1 above. Probably use index 0 as an example or add a comment at the end of the line saying that 1 is the array index.

File: public/apps/configuration/utils/combo-box-utils.tsx [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+ *   express or implied. See the License for the specific language governing
+ *   permissions and limitations under the License.
+ */
+
+import { EuiComboBoxOptionOption } from '@elastic/eui';
+import { Dispatch, SetStateAction } from 'react';
+import { StringRepresentable, curry } from 'lodash';
+import { appendElementToArray } from './array-state-utils';
+
+// Build option objects for EuiComboBox
```
Comment (tianleh) [inline public/apps/configuration/utils/combo-box-utils.tsx:21]: Minor: Do you mean a singular object? Probably call it "Build an option object"

File: public/apps/configuration/utils/combo-box-utils.tsx [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+export function stringToComboBoxOption(option: string): EuiComboBoxOptionOption {
+  return { label: option };
+}
+
+// Unbuild EuiComboBox option objects to string[]
+export function comboBoxOptionToString(option: EuiComboBoxOptionOption): string {
```
Comment (tianleh) [inline public/apps/configuration/utils/combo-box-utils.tsx:27]: Same here. I think you are handling singular case, not plural case. 
Unbuild an EuiComboBox option object to string

File: public/apps/configuration/utils/combo-box-utils.tsx [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  return option.label;
+}
+
+/**
+ * Convert a string to an combo box option and append to an array (or sub array indicating by path)
```
Comment (tianleh) [inline public/apps/configuration/utils/combo-box-utils.tsx:32]: Minor: an combo box -> a combo box

File: public/apps/configuration/utils/form-row.tsx [hunk 6]
```diff
@@ -13,28 +13,40 @@
  *   permissions and limitations under the License.
  */
 
+import { EuiFormRow, EuiLink, EuiText } from '@elastic/eui';
 import React from 'react';
-import { EuiFormRow, EuiText, EuiTitle } from '@elastic/eui';
```
Comment (tianleh) [inline public/apps/configuration/utils/form-row.tsx:17]: Just curious whether the location change is intentional or have been done by eslint?

Comment (yuxitang-amzn) [reply public/apps/configuration/utils/form-row.tsx:17]: Done by prettifier

File: public/apps/configuration/utils/combo-box-utils.tsx (hunk 3 above)
Comment (yuxitang-amzn) [reply public/apps/configuration/utils/combo-box-utils.tsx]: Done improvement.

File: public/apps/configuration/utils/combo-box-utils.tsx (hunk 5 above)
Comment (yuxitang-amzn) [reply public/apps/configuration/utils/combo-box-utils.tsx]: Done improvement.

File: public/apps/configuration/utils/combo-box-utils.tsx (hunk 4 above)
Comment (yuxitang-amzn) [reply public/apps/configuration/utils/combo-box-utils.tsx:27]: Done improvement.

File: public/apps/configuration/utils/array-state-utils.tsx (hunk 2 above)
Comment (yuxitang-amzn) [reply public/apps/configuration/utils/array-state-utils.tsx]: Done improvement.

File: public/apps/configuration/panels/role-edit.tsx (hunk 1 above)
Comment (yuxitang-amzn) [reply public/apps/configuration/panels/role-edit.tsx:17]: Agreed, didn't think it could be this large at very beginning.

**Pull request merged**